### PR TITLE
feat: swagger v2

### PR DESCRIPTION
**Pull request by Simperfit**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | yes if we remove nelmio
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #538
| License       | MIT
| Doc PR        | todo

Since NelmioApiDocBundle is not going to update to swagger v2, we will do it in api-platform.

This is WIP : 
- [x] Add tests (unit / behat) [ see https://travis-ci.org/api-platform/core/jobs/142226685#L4610]
- [x] Refactor, delete maximum of dupplication
- [x] look at scrutinizer

I have been doing that today, tested with SwaggerUI by changing the format of hydra / swagger

### CODE REVIEW

File: src/Bridge/Symfony/Bundle/DependencyInjection/ApiPlatformExtension.php [hunk 1]
```diff
@@ -110,14 +110,15 @@ public function load(array $configs, ContainerBuilder $container)
...
-     * Enables JSON-LD and Hydra support.
+     * Enables JSON-LD and Hydra and Swagger support.
      *
      * @param XmlFileLoader $loader
      */
     private function enableJsonLd(XmlFileLoader $loader)
     {
         $loader->load('jsonld.xml');
         $loader->load('hydra.xml');
+        $loader->load('swagger.xml');
```
Comment (dunglas) [inline src/Bridge/Symfony/Bundle/DependencyInjection/ApiPlatformExtension.php]: Swagger support is not related to JSON-LD support. You can import the file directly here: https://github.com/Simperfit/core/blob/119af86fdd0ba61090320d73055b3e87d50c2e2a/src/Bridge/Symfony/Bundle/DependencyInjection/ApiPlatformExtension.php#L84

File: src/Bridge/Symfony/Validator/Hydra/EventListener/HydraValidationExceptionListener.php [hunk 2]
```diff
@@ -21,7 +21,7 @@
  *
  * @author Kévin Dunglas <dunglas@gmail.com>
  */
-final class ValidationExceptionListener
+final class HydraValidationExceptionListener
```
Comment (dunglas) [inline src/Bridge/Symfony/Validator/Hydra/EventListener/HydraValidationExceptionListener.php]: Why this change?

File: src/Bridge/Symfony/Validator/Serializer/ConstraintViolationListNormalizer.php [hunk 3]
```diff
@@ -22,7 +22,7 @@
  */
 final class ConstraintViolationListNormalizer implements NormalizerInterface
 {
-    const FORMAT = 'hydra-error';
+    const FORMAT = ['swagger-error', 'hydra-error'];
```
Comment (dunglas) [inline src/Bridge/Symfony/Validator/Serializer/ConstraintViolationListNormalizer.php]: IMO Hydra and Swagger support should be 100% independent and the resulting documents shouldn't mix both formats.

File: src/Swagger/ApiDocumentationBuilder.php [hunk 4]
```diff
@@ -0,0 +1,293 @@
...
+
+        $doc['swagger'] = '2.0';
+        if ('' !== $this->title) {
+            $doc['info']['title'] = $this->title;
+        }
+
+        if ('' !== $this->description) {
+            $doc['info']['description'] = $this->description;
+        }
+        $doc['host'] = $_SERVER['HTTP_HOST'];
```
Comment (dunglas) [inline src/Swagger/ApiDocumentationBuilder.php]: Super-globals should never be accessed directly. This data would better be injected in the constructor.

File: src/Swagger/ApiDocumentationBuilder.php [hunk 5]
```diff
@@ -0,0 +1,293 @@
...
+        if (!$type) {
+            return;
+        }
+
+        if ($type->isCollection() && $collectionType = $type->getCollectionValueType()) {
+            $type = $collectionType;
+        }
+
+        switch ($type->getBuiltinType()) {
+            case 'string':
```
Comment (dunglas) [inline src/Swagger/ApiDocumentationBuilder.php]: Constants from the `Type` class should be used.

File: src/Swagger/ApiDocumentationBuilder.php [hunk 6]
```diff
@@ -0,0 +1,293 @@
...
+                return 'number';
+
+            case 'bool':
+                return 'boolean';
+
+            case 'object':
+                $className = $type->getClassName();
+
+                if ($className) {
+                    if ('DateTime' === $className) {
```
Comment (dunglas) [inline src/Swagger/ApiDocumentationBuilder.php]: `if (is_subclass_of($className, \DateTimeInterface::class)) {`

File: src/Swagger/ApiDocumentationBuilder.php [hunk 7]
```diff
@@ -0,0 +1,293 @@
...
+                        return 'string';
+                    }
+
+                    $className = $type->getClassName();
+                    if ($this->resourceClassResolver->isResourceClass($className)) {
+                        return ['$ref' => sprintf('#/definitions/%s', $this->resourceMetadataFactory->create($className)->getShortName())];
+                    }
```
Comment (dunglas) [inline src/Swagger/ApiDocumentationBuilder.php]: Is it ok to ignore classes that are not dates nor resources?

File: src/Swagger/EventListener/SwaggerLinkHeaderResponseListener.php [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+    }
+
+    /**
+     * Sends the Swagger header on each response.
+     *
+     * @param FilterResponseEvent $event
+     */
+    public function onKernelResponse(FilterResponseEvent $event)
+    {
+        if ('application/swagger' === $event->getResponse()->headers->get('Content-Type')) {
```
Comment (dunglas) [inline src/Swagger/EventListener/SwaggerLinkHeaderResponseListener.php]: Can't find anything about that in the spec. Where do it come from?

**Comment by dunglas** [comment]
First of all, thanks for working on this @Simperfit. It's very much appreciated.

Unless JSON-LD, HAL or JSON API, Swagger is just an API documentation format. Not all resources exposed by the API will be available in the "Swagger". There will be only one new URL returning the Swagger doc.

Basically, you can keep your `Swagger\ApiDocumentationBuilder` and add a new Swagger action (mapped to a `/swagger` path for instance) and you're all done. When someone will want to use Swagger UI, it will query this URL. Then Swagger UI will send JSON-LD requests to the API.

File: src/Swagger/EventListener/SwaggerLinkHeaderResponseListener.php (hunk 8 above)
Comment (Simperfit) [inline src/Swagger/EventListener/SwaggerLinkHeaderResponseListener.php]: It was just for testing purpose

File: src/Swagger/ApiDocumentationBuilder.php (hunk 7 above)
Comment (Simperfit) [inline src/Swagger/ApiDocumentationBuilder.php]: I guess yes because they are not definitions

**Comment by Simperfit** [comment]
@dunglas Comments taken in account. I've added a console command that dumps the documentation

File: src/Bridge/Symfony/Bundle/Command/SwaggerCommand.php [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+use Symfony\Component\Console\Command\Command;
+use Symfony\Component\Console\Input\InputInterface;
+use Symfony\Component\Console\Output\OutputInterface;
+
+/**
+ * Console command to dump Swagger API documentations.
+ *
+ * @author Amrouche Hamza <hamza.simperfit@gmail.com>
+ */
+class SwaggerCommand extends Command
```
Comment (dunglas) [inline src/Bridge/Symfony/Bundle/Command/SwaggerCommand.php]: Can you make this class `final`?

File: src/Bridge/Symfony/Bundle/Command/SwaggerCommand.php [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+{
+    /**
+     * @var ApiDocumentationBuilder
```
Comment (dunglas) [inline src/Bridge/Symfony/Bundle/Command/SwaggerCommand.php]: Useless dockblock, should be removed.

File: src/Bridge/Symfony/Bundle/Command/SwaggerCommand.php [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+    protected $apiDocumentationBuilder;
+
+    public function __construct(ApiDocumentationBuilder $apiDocumentationBuilder)
+    {
+        $this->apiDocumentationBuilder = $apiDocumentationBuilder;
+
+        parent::__construct();
+    }
+
+    protected function configure()
```
Comment (dunglas) [inline src/Bridge/Symfony/Bundle/Command/SwaggerCommand.php]:
```php
/**
  * {@inheritdoc}
  */
```

missing

File: src/Bridge/Symfony/Bundle/Command/SwaggerCommand.php [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+    {
+        $this
+            ->setName('api:swagger:export')
+            ->setDescription('Export a beautiful json of swagger api');
```
Comment (dunglas) [inline src/Bridge/Symfony/Bundle/Command/SwaggerCommand.php]: `Dump the Swagger 2.0 (OpenAPI) documentation`?

File: src/Bridge/Symfony/Bundle/Command/SwaggerCommand.php [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+    }
+
+    protected function execute(InputInterface $input, OutputInterface $output)
```
Comment (dunglas) [inline src/Bridge/Symfony/Bundle/Command/SwaggerCommand.php]:
```php
/**
 * {@inheritdoc}
 */
```

missing

File: src/Bridge/Symfony/Bundle/Resources/config/hydra.xml [hunk 14]
```diff
@@ -52,7 +52,7 @@
             <tag name="serializer.normalizer" priority="50" />
         </service>
 
-        <service id="api_platform.hydra.normalizer.constraint_violation_list" class="ApiPlatform\Core\Bridge\Symfony\Validator\Hydra\Serializer\ConstraintViolationListNormalizer" public="false">
+        <service id="api_platform.hydra.normalizer.constraint_violation_list" class="ApiPlatform\Core\Bridge\Symfony\Validator\Serializer\ConstraintViolationListNormalizer" public="false">
```
Comment (dunglas) [inline src/Bridge/Symfony/Bundle/Resources/config/hydra.xml]: Should be reverted.

File: src/Bridge/Symfony/Bundle/Resources/config/routing/swagger.xml [hunk 15]
```diff
@@ -0,0 +1,11 @@
+<?xml version="1.0" encoding="UTF-8" ?>
+
+<routes xmlns="http://symfony.com/schema/routing"
+        xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+        xsi:schemaLocation="http://symfony.com/schema/routing
+        http://symfony.com/schema/routing/routing-1.0.xsd">
+
+    <route id="api_swagger_vocab" path="/swagger.json">
```
Comment (dunglas) [inline src/Bridge/Symfony/Bundle/Resources/config/routing/swagger.xml]: `<route id="api_swagger" path="/swagger">`?

File: src/Bridge/Symfony/Validator/Serializer/ConstraintViolationListNormalizer.php [hunk 16]
```diff
@@ -9,7 +9,7 @@
  * file that was distributed with this source code.
  */
 
-namespace ApiPlatform\Core\Bridge\Symfony\Validator\Hydra\Serializer;
+namespace ApiPlatform\Core\Bridge\Symfony\Validator\Serializer;
```
Comment (dunglas) [inline src/Bridge/Symfony/Validator/Serializer/ConstraintViolationListNormalizer.php]: All changes in this file should be reverted.

File: src/Bridge/Symfony/Validator/Swagger/EventListener/SwaggerValidationExceptionListener.php [hunk 17]
```diff
@@ -0,0 +1,49 @@
...
+/*
+ * This file is part of the API Platform project.
+ *
+ * (c) Kévin Dunglas <dunglas@gmail.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace ApiPlatform\Core\Bridge\Symfony\Validator\Swagger\EventListener;
```
Comment (dunglas) [inline src/Bridge/Symfony/Validator/Swagger/EventListener/SwaggerValidationExceptionListener.php]: Changes in this file must be reverted too.

File: src/Hydra/Action/DocumentationAction.php [hunk 18]
```diff
@@ -11,7 +11,7 @@
 
 namespace ApiPlatform\Core\Hydra\Action;
 
-use ApiPlatform\Core\Hydra\ApiDocumentationBuilderInterface;
+use ApiPlatform\Core\Util\ApiDocumentationBuilderInterface;
```
Comment (dunglas) [inline src/Hydra/Action/DocumentationAction.php]: Useless, should be reverted.

File: src/Hydra/ApiDocumentationBuilder.php [hunk 19]
```diff
@@ -21,6 +21,8 @@
 use ApiPlatform\Core\Metadata\Resource\Factory\ResourceMetadataFactoryInterface;
 use ApiPlatform\Core\Metadata\Resource\Factory\ResourceNameCollectionFactoryInterface;
 use ApiPlatform\Core\Metadata\Resource\ResourceMetadata;
+use ApiPlatform\Core\Util\ApiDocumentationBuilderInterface;
```
Comment (dunglas) [inline src/Hydra/ApiDocumentationBuilder.php]: Useless, should be reverted.

File: src/JsonLd/Serializer/ItemNormalizer.php [hunk 20]
```diff
@@ -97,8 +97,10 @@ public function normalize($object, $format = null, array $context = [])
             return $rawData;
         }
 
-        $data['@id'] = $this->iriConverter->getIriFromItem($object);
-        $data['@type'] = ($iri = $resourceMetadata->getIri()) ? $iri : $resourceMetadata->getShortName();
+        if ($format === self::FORMAT) {
```
Comment (dunglas) [inline src/JsonLd/Serializer/ItemNormalizer.php]: Should be reverted.

File: src/Swagger/Action/DocumentationAction.php [hunk 21]
```diff
@@ -0,0 +1,44 @@
...
+    /**
+     * Gets API doc.
+     *
+     * @param Request $request
+     *
+     * @return array
+     */
+    public function __invoke(Request $request)
+    {
+        $request->attributes->set('_api_format', 'jsonld');
```
Comment (dunglas) [inline src/Swagger/Action/DocumentationAction.php]: Will change in #590

File: src/Swagger/ApiDocumentationBuilder.php [hunk 22]
```diff
@@ -0,0 +1,291 @@
...
+use ApiPlatform\Core\Metadata\Resource\ResourceMetadata;
+use ApiPlatform\Core\Util\ApiDocumentationBuilderInterface;
+use Symfony\Component\PropertyInfo\Type;
+
+/**
+ * Creates a machine readable Swagger API documentation.
+ *
+ * @author Kévin Dunglas <dunglas@gmail.com>
+ */
+final class ApiDocumentationBuilder implements ApiDocumentationBuilderInterface
```
Comment (dunglas) [inline src/Swagger/ApiDocumentationBuilder.php]: Does it really worth it to introduce a common API for the Hydra and the Swagger doc?

File: src/Hydra/ApiDocumentationBuilder.php (hunk 19 above)
Comment (dunglas) [inline src/Hydra/ApiDocumentationBuilder.php]: Or an `implements` statement is missing.

File: src/Util/ApiDocumentationBuilderInterface.php [hunk 23]
```diff
@@ -9,17 +9,17 @@
  * file that was distributed with this source code.
  */
 
-namespace ApiPlatform\Core\Hydra;
+namespace ApiPlatform\Core\Util;
```
Comment (dunglas) [inline src/Util/ApiDocumentationBuilderInterface.php]: OK why not keeping this command interface, but:

* Should be move to a `Documentation` namespace
* Can be used to have a common controller for both docs (and an optional `?format=swagger|hydra` for instance)

**Comment by dunglas** [comment]
When you'll write tests, you can use https://hub.docker.com/r/swaggerapi/swagger-validator/ to validate the generated Swagger file of the test app in Travis.

File: features/bootstrap/SwaggerContext.php [hunk 24]
```diff
@@ -0,0 +1,275 @@
...
+     *
+     * @param BeforeScenarioScope $scope
+     *
+     * @BeforeScenario
+     */
+    public function gatherContexts(BeforeScenarioScope $scope)
+    {
+        /** @var InitializedContextEnvironment $environment */
+        $environment = $scope->getEnvironment();
+        $this->restContext = $environment->getContext('Sanpi\Behatch\Context\RestContext');
```
Comment (theofidry) [inline features/bootstrap/SwaggerContext.php]: `Sanpi\Behatch\Context\RestContext::class`?

File: features/bootstrap/SwaggerContext.php [hunk 25]
```diff
@@ -0,0 +1,275 @@
...
+     * Gets an operation by its method name.
+     *
+     * @param string $className
+     * @param string $method
+     *
+     * @throws Exception
+     *
+     * @return array
+     */
+    private function getOperation($method, $className)
```
Comment (theofidry) [inline features/bootstrap/SwaggerContext.php]: as we are in php7 we can force the typehint everywhere and remove useless phpdoc

File: src/Bridge/Symfony/Bundle/Command/SwaggerCommand.php [hunk 26]
```diff
@@ -0,0 +1,54 @@
...
+use Symfony\Component\Console\Output\OutputInterface;
+
+/**
+ * Console command to dump Swagger API documentations.
+ *
+ * @author Amrouche Hamza <hamza.simperfit@gmail.com>
+ */
+final class SwaggerCommand extends Command
+{
+    protected $apiDocumentationBuilder;
```
Comment (theofidry) [inline src/Bridge/Symfony/Bundle/Command/SwaggerCommand.php]: `private`

**Comment by theofidry** [comment]
@Simperfit couldn't check it in details, but I left some notes here and there and you have some Scrutinizer warnings :)

File: src/Bridge/Symfony/Bundle/Command/SwaggerCommand.php (hunk 26 above)
Comment (Simperfit) [inline src/Bridge/Symfony/Bundle/Command/SwaggerCommand.php]: I cannot make the function private since its a command

Comment (dunglas) [inline src/Bridge/Symfony/Bundle/Command/SwaggerCommand.php]: Why? It's a Symfony limitation?

File: src/Bridge/Symfony/Bundle/Command/SwaggerCommand.php [hunk 27]
```diff
@@ -0,0 +1,54 @@
...
+
+    public function __construct(ApiDocumentationBuilder $apiDocumentationBuilder)
+    {
+        $this->apiDocumentationBuilder = $apiDocumentationBuilder;
+
+        parent::__construct();
```
Comment (dunglas) [inline src/Bridge/Symfony/Bundle/Command/SwaggerCommand.php]: Call the constructor before setting the property.

File: src/Bridge/Symfony/Bundle/Command/SwaggerCommand.php (hunk 26 above)
Comment (Simperfit) [inline src/Bridge/Symfony/Bundle/Command/SwaggerCommand.php]: Yes it is

File: features/bootstrap/SwaggerContext.php [hunk 28]
```diff
@@ -0,0 +1,260 @@
...
+
+use Behat\Behat\Context\Context;
+use Behat\Behat\Context\Environment\InitializedContextEnvironment;
+use Behat\Behat\Hook\Scope\BeforeScenarioScope;
+use Symfony\Component\PropertyAccess\PropertyAccess;
+
+class SwaggerContext implements Context
+{
+    /**
+     * @var PropertyAccessor
```
Comment (dunglas) [inline features/bootstrap/SwaggerContext.php]: Useless docblock

File: features/bootstrap/SwaggerContext.php [hunk 29]
```diff
@@ -0,0 +1,260 @@
...
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+use Behat\Behat\Context\Context;
+use Behat\Behat\Context\Environment\InitializedContextEnvironment;
+use Behat\Behat\Hook\Scope\BeforeScenarioScope;
+use Symfony\Component\PropertyAccess\PropertyAccess;
+
+class SwaggerContext implements Context
```
Comment (dunglas) [inline features/bootstrap/SwaggerContext.php]: Can you make this class `final`?

File: features/bootstrap/SwaggerContext.php [hunk 30]
```diff
@@ -0,0 +1,260 @@
...
+     *
+     * @param BeforeScenarioScope $scope
+     *
+     * @BeforeScenario
+     */
+    public function gatherContexts(BeforeScenarioScope $scope)
+    {
+        /** @var InitializedContextEnvironment $environment */
+        $environment = $scope->getEnvironment();
+        $this->restContext = $environment->getContext(Sanpi\Behatch\Context\RestContext::class);
```
Comment (dunglas) [inline features/bootstrap/SwaggerContext.php]: Can you add a `use` statement instead of using the FQN?

File: src/Bridge/Symfony/Routing/ApiLoader.php [hunk 31]
```diff
@@ -60,6 +60,7 @@ public function load($data, $type = null)
 
         $routeCollection->addCollection($this->fileLoader->load('jsonld.xml'));
         $routeCollection->addCollection($this->fileLoader->load('hydra.xml'));
+        $routeCollection->addCollection($this->fileLoader->load('swagger.xml'));
```
Comment (dunglas) [inline src/Bridge/Symfony/Routing/ApiLoader.php]: Should be loaded only if the config flag is set to `true`.

File: src/Bridge/Symfony/Validator/Hydra/Serializer/ConstraintViolationListNormalizer.php [hunk 32]
```diff
@@ -22,7 +22,7 @@
  */
 final class ConstraintViolationListNormalizer implements NormalizerInterface
 {
-    const FORMAT = 'hydra-error';
+    const FORMAT = ['hydra-error'];
```
Comment (dunglas) [inline src/Bridge/Symfony/Validator/Hydra/Serializer/ConstraintViolationListNormalizer.php]: Should be reverted.

Comment (dunglas) [inline src/Bridge/Symfony/Validator/Hydra/Serializer/ConstraintViolationListNormalizer.php]: Same for all changes in this file.

File: src/JsonLd/Serializer/ItemNormalizer.php [hunk 33]
```diff
@@ -100,6 +100,7 @@ public function normalize($object, $format = null, array $context = [])
         $data['@id'] = $this->iriConverter->getIriFromItem($object);
         $data['@type'] = ($iri = $resourceMetadata->getIri()) ? $iri : $resourceMetadata->getShortName();
 
+
```
Comment (dunglas) [inline src/JsonLd/Serializer/ItemNormalizer.php]: Should be reverted.

File: src/Swagger/ApiDocumentationBuilder.php [hunk 34]
```diff
@@ -0,0 +1,343 @@
...
+use ApiPlatform\Core\Metadata\Property\PropertyMetadata;
+use ApiPlatform\Core\Metadata\Resource\Factory\ResourceMetadataFactoryInterface;
+use ApiPlatform\Core\Metadata\Resource\Factory\ResourceNameCollectionFactoryInterface;
+use ApiPlatform\Core\Metadata\Resource\ResourceMetadata;
+use Symfony\Component\PropertyInfo\Type;
+
+/**
+ * Creates a machine readable Swagger API documentation.
+ *
+ * @author Kévin Dunglas <dunglas@gmail.com>
```
Comment (dunglas) [inline src/Swagger/ApiDocumentationBuilder.php]: You can credit yourself here :)

File: src/Swagger/ApiDocumentationBuilder.php [hunk 35]
```diff
@@ -0,0 +1,343 @@
...
+                    $property[$propertyName] = $range;
+                }
+
+                $required = [];
+
+                if ($propertyMetadata->isRequired()) {
+                    $required = array_merge($required, [$propertyName]);
+                }
+
+                if (0 !== count($required)) {
```
Comment (dunglas) [inline src/Swagger/ApiDocumentationBuilder.php]: `!empty` (micro optimization)

File: src/Swagger/ApiDocumentationBuilder.php [hunk 36]
```diff
@@ -0,0 +1,343 @@
...
+                foreach ($operations as $operationName => $itemOperation) {
+                    $swaggerOperation = $this->getSwaggerOperation($resourceClass, $resourceMetadata, $operationName, $itemOperation, $prefixedShortName, false);
+                    $itemOperations['operation'] = array_merge($itemOperations['operation'], $swaggerOperation);
+                }
+            }
+
+            try {
+                $resourceClassIri = $this->iriConverter->getIriFromResourceClass($resourceClass);
+            } catch (InvalidArgumentException $e) {
+                $resourceClassIri = '/nopaths';
```
Comment (dunglas) [inline src/Swagger/ApiDocumentationBuilder.php]: What is the reason of this?

File: src/Swagger/ApiDocumentationBuilder.php [hunk 37]
```diff
@@ -0,0 +1,343 @@
...
+            }
+            $resourceClassIri .= '/{id}';
+
+            $itemOperationsDocs[$resourceClassIri] = $itemOperations['operation'];
+            $classes[] = $class;
+        }
+
+        $doc['swagger'] = '2.0';
```
Comment (dunglas) [inline src/Swagger/ApiDocumentationBuilder.php]: What about using a class constant?

File: src/Swagger/ApiDocumentationBuilder.php [hunk 38]
```diff
@@ -0,0 +1,343 @@
...
+        }
+
+        if ('' !== $this->description) {
+            $doc['info']['description'] = $this->description;
+        }
+        $doc['info']['version'] = $this->version ?? '0.0.0';
+        $doc['host'] = $this->host;
+        $doc['basePath'] = $this->urlGenerator->generate('api_jsonld_entrypoint');
+        $doc['definitions'] = $properties;
+        $doc['externalDocs'] = ['description' => 'Find more about api-platform', 'url' => 'docs'];
```
Comment (dunglas) [inline src/Swagger/ApiDocumentationBuilder.php]: `$doc['externalDocs'] = ['description' => 'Find more about API Platform', 'url' => 'https://api-platform.com'];`?

File: features/bootstrap/SwaggerContext.php (hunk 29 above)
Comment (theofidry) [inline features/bootstrap/SwaggerContext.php]: @dunglas is it necessary? It's a context so it shouldn't be part of the classmap if installed with dev dependencies. We're not writing our test cases as final either

File: src/Swagger/ApiDocumentationBuilder.php [hunk 39]
```diff
@@ -0,0 +1,343 @@
...
+            $method = $this->operationMethodResolver->getCollectionOperationMethod($resourceClass, $operationName);
+        } else {
+            $method = $this->operationMethodResolver->getItemOperationMethod($resourceClass, $operationName);
+        }
+        $methodSwagger = strtolower($method);
+        $swaggerOperation = $operation['swagger_context'] ?? [];
+        $shortName = $resourceMetadata->getShortName();
+        $swaggerOperation[$methodSwagger] = [];
+        $swaggerOperation[$methodSwagger]['tags'] = [$shortName];
+        $swaggerOperation[$methodSwagger]['produces'] = ['application/json'];
```
Comment (dunglas) [inline src/Swagger/ApiDocumentationBuilder.php]: 'application/ld+json`?

File: src/Swagger/ApiDocumentationBuilder.php [hunk 40]
```diff
@@ -0,0 +1,343 @@
...
+                        'schema' => [
+                            '$ref' => sprintf('#/definitions/%s', $shortName),
+                        ],
+                    ], ];
+                }
+
+                $swaggerOperation[$methodSwagger]['responses'] = [
+                    '200' => ['description' => 'Valid ID'],
+                ];
+
```
Comment (dunglas) [inline src/Swagger/ApiDocumentationBuilder.php]: Extra line to remove.

File: src/Swagger/ApiDocumentationBuilder.php [hunk 41]
```diff
@@ -0,0 +1,343 @@
...
+            break;
+            case 'DELETE':
```
Comment (dunglas) [inline src/Swagger/ApiDocumentationBuilder.php]: Add a blank line before.

File: .travis.yml [hunk 42]
```diff
@@ -22,10 +22,13 @@ before_install:
     - if [ "$TRAVIS_PHP_VERSION" != "hhvm" ]; then phpenv config-rm xdebug.ini; fi;
     - phpunit --self-update
     - if [ "$SYMFONY_VERSION" != "" ]; then composer require "symfony/symfony:${SYMFONY_VERSION}" --no-update; fi;
+    - npm install -g swagger-cli
```
Comment (theofidry) [inline .travis.yml]: outch, IMO it would be better to have it in a `packages.json` and shrinkwrap it. `node_modules` could then be cached like it is for composer

File: features/bootstrap/SwaggerContext.php [hunk 43]
```diff
@@ -0,0 +1,260 @@
...
+    public function assertPropertyNotExist($propertyName, $className)
+    {
+        try {
+            $this->getProperty($propertyName, $className);
+
+            throw new \PHPUnit_Framework_ExpectationFailedException(
+                sprintf('The property "%s" for the class "%s" exist.', $propertyName, $className)
+            );
+        } catch (\Exception $exception) {
+            // an exception must be catched
```
Comment (theofidry) [inline features/bootstrap/SwaggerContext.php]: ?

File: features/bootstrap/SwaggerContext.php [hunk 44]
```diff
@@ -0,0 +1,260 @@
...
+        $properties = $this->getProperties($className);
+        $propertyInfos = null;
+        foreach ($properties as $classPropertyName => $property) {
+            if ($classPropertyName === $propertyName) {
+                $propertyInfos = $property;
+            }
+        }
+
+        if (empty($propertyInfos)) {
+            throw new \Exception(sprintf('Property "%s" of class "%s" does\'nt exist', $propertyName, $className));
```
Comment (theofidry) [inline features/bootstrap/SwaggerContext.php]: `does not`

**Comment by dunglas** [comment]
I left some minor comments but it looks good to me. What do you think @api-platform/core-team I want it in the first beta too!

File: features/bootstrap/SwaggerContext.php [hunk 45]
```diff
@@ -0,0 +1,260 @@
...
+     *
+     * @return array | stdClass
+     */
+    private function getProperty(string $propertyName, string $className) : stdClass
+    {
+        $properties = $this->getProperties($className);
+        $propertyInfos = null;
+        foreach ($properties as $classPropertyName => $property) {
+            if ($classPropertyName === $propertyName) {
+                $propertyInfos = $property;
```
Comment (theofidry) [inline features/bootstrap/SwaggerContext.php]: do we have to iterate over all the properties? Can't we return as soon as we got one?

File: features/bootstrap/SwaggerContext.php [hunk 46]
```diff
@@ -0,0 +1,260 @@
...
+
+    /**
+     * Gets an operation by its method name.
+     *
+     * @param string $className
+     * @param string $method
+     *
+     * @throws Exception
+     *
+     * @return array | stdClass
```
Comment (theofidry) [inline features/bootstrap/SwaggerContext.php]: IMO everything except the `Gets an operation by its method name.` can be removed. As the context is not meant to be reused, indicating that an exception is thrown is useless

Comment (theofidry) [inline features/bootstrap/SwaggerContext.php]: you also have a return typehint missmatch

File: features/bootstrap/SwaggerContext.php [hunk 47]
```diff
@@ -0,0 +1,260 @@
...
+     */
+    private function getOperations(string $className) : stdClass
+    {
+        $classInfos = $this->getClassInfos($className);
+
+        return isset($classInfos) ? $classInfos : [];
```
Comment (theofidry) [inline features/bootstrap/SwaggerContext.php]: don't we have something better than `isset` here? `$classInfo` is an array right? so we can use `empty` instead

File: .travis.yml (hunk 42 above)
Comment (dunglas) [inline .travis.yml]: IMO it's better as is. Having a `packages.json` in the repository (that doesn't use any NPM dependency) has no sense and it's good to test the Swagger support with the latest available version of swagger-cli (we also install the last version of composer for instance).

Comment (theofidry) [inline .travis.yml]: >Having a packages.json in the repository (that doesn't use any NPM dependency) has no sense

Well, with this we are. If we don't add it it's a hidden project dependency. Not required unless you are developing on this bit, but hidden still. Having a `packages.json` does not force you to do a `npm install` when you want to get started with the project, but at least it avoids people to have to look at Travis to figure out that they need `swagger-cli`.

>it's good to test the Swagger support with the latest available version of swagger-cli

In that case yeah we can drop the `npm-shrinkwrap.json` thing, which makes it even easier as unlike composer, shrinkwrap depends on the local node version installed

Comment (dunglas) [inline .travis.yml]: They don't need it to develop. It's just an extra check we do in the CI to be sure that the output of the Swagger generator is correct.
If we used a go validator (there is a goswagger is able to generate a Swagger file), we would not have versioned a go setup in the repo just to use this program.

Comment (theofidry) [inline .travis.yml]: fair enough

**Comment by dunglas** [comment]
Can you also add a unit test for the Swagger generator?

File: features/bootstrap/SwaggerContext.php (hunk 29 above)
Comment (dunglas) [inline features/bootstrap/SwaggerContext.php]: Ok I agree with @theofidry

File: features/bootstrap/SwaggerContext.php (hunk 43 above)
Comment (dunglas) [inline features/bootstrap/SwaggerContext.php]: I guess it's to respect PSR-2 that forbid empty catch without comment. Maybe a better message can be found :-)

File: src/Swagger/ApiDocumentationBuilder.php (hunk 36 above)
Comment (Simperfit) [inline src/Swagger/ApiDocumentationBuilder.php]: Because in the tests `NoCollectionDummy` does throw an error in here.

File: src/Swagger/ApiDocumentationBuilder.php (hunk 37 above)
Comment (Simperfit) [inline src/Swagger/ApiDocumentationBuilder.php]: I was thinking of a parameters in the service too ?

Comment (dunglas) [inline src/Swagger/ApiDocumentationBuilder.php]: IMO a constant is better because the version cannot be changed without patching the code.

File: features/bootstrap/SwaggerContext.php (hunk 46 above)
Comment (Simperfit) [inline features/bootstrap/SwaggerContext.php]: done

**Comment by Simperfit** [comment]
@dunglas I asserted the Json, what do I test ? Like in behat ?

**Comment by dunglas** [comment]
@Simperfit a PHPUnit test on the class itself would be the best. Can you resolve conflicts too?

**Comment by Simperfit** [comment]
@dunglas On the console or on the ApiDocumentationBuilder ?
I have rebased

**Comment by dunglas** [comment]
On the builder. We try to have unit tests for every new classes.

**Comment by dunglas** [comment]
Thank you @Simperfit
